src/renderers/shared/fiber/ReactFiberUpdateQueue.js

```diff
@@ -12,100 +12,473 @@
+    queue = {
+      first: null,
+      last: null,
+      isProcessing: false,
+    };
+  } else {
+    queue = {
+      first: null,
+      last: null,
+    };
+  }
+
+  fiber.updateQueue = queue;
+  return queue;
+}
+
+// Clones an update queue from a source fiber onto its alternate.
+function cloneUpdateQueue(alt : Fiber, fiber : Fiber) : UpdateQueue | null {
+  const sourceQueue = fiber.updateQueue;
+  if (!sourceQueue) {
+    // The source fiber does not have an update queue.
+    alt.updateQueue = null;
+    return null;
+  }
+  // If the alternate already has a queue, reuse the previous object.
+  const altQueue = alt.updateQueue || {};
+  altQueue.first = sourceQueue.first;
+  altQueue.last = sourceQueue.last;
+  alt.updateQueue = altQueue;
+  return altQueue;
+}
+exports.cloneUpdateQueue = cloneUpdateQueue;
+
+function cloneUpdate(update : Update) : Update {
+  return {
+    priorityLevel: update.priorityLevel,
+    partialState: update.partialState,
+    callback: update.callback,
+    isReplace: update.isReplace,
+    isForced: update.isForced,
+    next: null,
+  };
+}
+
+function insertUpdateIntoQueue(queue, update, insertAfter, insertBefore) {
+  if (insertAfter) {
+    insertAfter.next = update;
+  } else {
+    // This is the first item in the queue.
+    update.next = queue.first;
+    queue.first = update;
+  }
+
+  if (insertBefore) {
+    update.next = insertBefore;
+  } else {
+    // This is the last item in the queue.
+    queue.last = update;
+  }
+}
+
+// The work-in-progress queue is a subset of the current queue (if it exists).
+// We need to insert the incoming update into both lists. However, it's possible
+// that the correct position in one list will be different from the position in
+// the other. Consider the following case:
+//
+//     Current:             3-5-6
+//     Work-in-progress:        6
+//
+// Then we receive an update with priority 4 and insert it into each list:
+//
+//     Current:             3-4-5-6
+//     Work-in-progress:        4-6
+//
+// In the current queue, the new update's `next` pointer points to the update
+// with priority 5. But in the work-in-progress queue, the pointer points to the
+// update with priority 6. Because these two queues share the same persistent
+// data structure, this won't do. (This can only happen when the incoming update
+// has higher priority than all the updates in the work-in-progress queue.)
+//
+// To solve this, in the case where the incoming update needs to be inserted
+// into two different positions, we'll make a clone of the update and insert
+// each copy into a separate queue. This forks the list while maintaining a
+// persistent stucture, because the update that is added to the work-in-progress
+// is always added to the front of the list.
+//
+// However, if incoming update is inserted into the same position of both lists,
+// we shouldn't make a copy.
+
+function insertUpdate(fiber : Fiber, update : Update, methodName : ?string) : void {
```

Comment:
I suppose you could move the error out of this function. It's a pretty simple conditional to replicate to both methods. That would be nicer.